[WIP] Multiple Service CIDRs

aojea opened the pull request:
/kind feature
/kind api-change

#### What this PR does / why we need it:

- Implements [KEP 1889 - Multiple Service CIDRs](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1880-multiple-service-cidrs) , so users can configure the ServiceCIDRs via the API
- Fixes the ServiceCIDR size limitation 2^16 for IPv6. Current limitation is 2^64-1 to avoid dealing with uint64 overflows
- Fixes the bootstrap and "kubernetes.default" Service creation when multiple apiservers use different flags for the service-cidr ranges
- ....


#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

```
$ kubectl get servicecidr
NAME                       IPV4           IPV6
kubernetes-default-hf5wd   10.96.0.0/16   <none>

$ kubectl get services -A
NAMESPACE     NAME         TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)                  AGE
default       kubernetes   ClusterIP   10.96.0.1    <none>        443/TCP                  5m42s
kube-system   kube-dns     ClusterIP   10.96.0.10   <none>        53/UDP,53/TCP,9153/TCP   5m36s

$ kubectl get ipaddresses
NAME         PARENTREF
10.96.0.1    services/default/kubernetes
10.96.0.10   services/kube-system/kube-dns
```


#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1880-multiple-service-cidrs

```

leilajal commented:
/triage accepted

liggitt commented on api/api-rules/violation_exceptions.list:
```diff
@@ -393,6 +393,9 @@ API rule violation: names_match,k8s.io/api/core/v1,RBDVolumeSource,RBDPool
 API rule violation: names_match,k8s.io/api/core/v1,RBDVolumeSource,RadosUser
 API rule violation: names_match,k8s.io/api/core/v1,VolumeSource,CephFS
 API rule violation: names_match,k8s.io/api/core/v1,VolumeSource,StorageOS
+API rule violation: names_match,k8s.io/api/networking/v1alpha1,ParentReference,Resource
+API rule violation: names_match,k8s.io/api/networking/v1alpha1,ServiceCIDRConfigSpec,IPv4CIDR
+API rule violation: names_match,k8s.io/api/networking/v1alpha1,ServiceCIDRConfigSpec,IPv6CIDR
```
we shouldn't add new violations, right?

aojea replied:
I'm just trying to make it work :smile: , there are a lot of things here that are magic to me,

aojea committed:
external API

add new label to public API

add new label to the public API

move constant from its own file to the types.go on external API

remove servicecidrm in external API

finalizer to external API

finalizer to external AIP

aojea committed:
add feature gate

rename feature

TEST enable serviceCIDR feature by default

aojea committed:
hack import aliases

aojea committed:
test integration etcd data

aojea committed:
api rules violation.list

aojea committed:
add apis to apiserver storage

aojea committed:
kubectl describe

fix kubectl describe

fix kubectl describe test

aojea committed:
make update

aojea committed:
add internal API

aojea committed:
add internal API validation

aojea committed:
enable networkingv1alpha1 ServiceCIDR and IPAddress

on the storage

aojea committed:
Add ServiceCIDR to the REST Registry

aojea committed:
Add IPAddress API to the REST registry

aojea committed:
add printers MISSING TESTS

aojea committed:
refactor current ipallocator

- rename files to match the allocator backend
- use t.Run for tests and cover large ranges
- add benchmarks

goos: linux
goarch: amd64

pkg: k8s.io/kubernetes/pkg/registry/core/service/ipallocator
cpu: Intel(R) Xeon(R) CPU E5-2678 v3 @ 2.50GHz
BenchmarkAllocateNextIPv4Size1048574
BenchmarkAllocateNextIPv4Size1048574-24    	 1517683
7373 ns/op	     135 B/op	       8 allocs/op
BenchmarkAllocateNextIPv6Size65535
BenchmarkAllocateNextIPv6Size65535-24      	 5607438
193.9 ns/op	      18 B/op	       2 allocs/op
PASS

aojea committed:
check that thebitmap ip allocator satisfies the interface

aojea commented:
ping @thockin for the API review on the first 2 commits

aojea committed:
add integration test

integration test

test integration fix double import

rename feature test integration

aojea committed:
WIP e2e skaffolding

aojea committed:
TEST enable networking v1alpha1 by default

aojea committed:
initialize allocator on the controlplane instance

The allocators for Services ClusterIPs and NodePorts requires to
share information between the controlplane and the Rest storage.

Instead of initializing them on the rest storage, and sharing them
with the controlplane instance, do the other way around, initialize
them on the controlplane instance and share them with the rest storage.

aojea committed:
build the service allocators in its own function

aojea committed:
Controller-manager ServiceCIDR and IPAddress controller

Controller that reconciles Services, ServiceCIDRs and
IPAddresses.

It lives in the controller-manager

aojea committed:
add ip radix-tree as util package

fix iptree

add tests iptree

document iptree

improve iptree

iptree

iptree

iptree

fix iptree bug

remove unused nodes

aojea committed:
add the new service allocators to the controlplane instance

aojea committed:
New ServiceCIDRs controlplane controller

Controller on the controlplane instance that handles the creation of the
default ServiceCIDR, it has to keep compatibility with the actual
apiserver flags and handle different configurations.

aojea committed:
enable the new ClusterIP allocator on the controlplane instance

aojea committed:
bootstrap create default namespace

aojea committed:
controller bootstrap auth policy for controller manager

aojea committed:
enforce IPAddress reference field in validation

aojea commented:
lovely, it passes CI e2e :smile:

aojea committed:
debug test integration

aojea commented on pkg/features/kube_features.go:
```diff
@@ -735,7 +735,14 @@ const (
+	// alpha: v1.25
+	//
+	// Allow to dynamically configure Service IP ranges
+	ServiceCIDR featuregate.Feature = "ServiceCIDR"
```
s/ServiceCIDR/MultipleServiceCIDR/

aojea committed:
add new clusterIP allocator based on the new APIs

aojea committed:
rename bitmap allocator to range allocator

aojea committed:
enable the new ClusterIP allocator based on the feature gate

aojea committed:
TEST enable MultiCIDRServiceAllocator feature gate

aojea commented on pkg/features/kube_features.go:
```diff
@@ -775,7 +782,7 @@ const (
 
 	// owner: @aojea
-	// kep: https://kep.k8s.io/3070
+	// kep: http://kep.k8s.io/3070
```
https

aojea commented on pkg/registry/networking/servicecidr/strategy_test.go:
```diff
@@ -0,0 +1,17 @@
+/*
+Copyright 2021 The Kubernetes Authors.
```
2022

aojea commented:
back on track,

aojea commented on pkg/apis/networking/validation/validation.go:
```diff
@@ -741,3 +743,129 @@ func validateClusterCIDRUpdateSpec(update, old *networking.ClusterCIDRSpec, fldP
+	ip, err := netip.ParseAddr(name)
+	if err != nil {
+		errs = append(errs, err.Error())
+	} else if ip.String() != name {
```
Ben: we should check the contract if this always return canonical

aojea commented on pkg/registry/core/service/ipallocator/service_cidrs.go:
```diff
@@ -0,0 +1,580 @@
+	// allocate an IP from one of the available ServiceCIDRs
+	a.muTree.Lock()
+	cidrs := a.tree.Parents()
+	a.muTree.Unlock()
```
should we lock the entire operation?

aojea commented on pkg/registry/core/service/ipallocator/service_cidrs.go:
```diff
@@ -0,0 +1,580 @@
+		if err != nil {
+			return nil, err
+		}
+		// find an empty address within the range
+		if int64(len(ipAddresses)) >= max {
+			continue
+		}
```
this was only valid when we used the label to assign ips to cidrs

aojea committed:
CODE REFACTORING TO PLUG THE NEW APIS ON THE APISERVER

Change-Id: I681eb8906900193f14c87666fdfe527d45002fdb

aojea committed:
START API REVIEW HERE

Change-Id: I9bf43cf815840bb80d3d2db3ae8f762afd62b711

aojea committed:
END API REVIEW HERE

Change-Id: Ief91bdfa4ac76d9183297721463965f86cd53452

robscott commented:
/cc

dims commented:
This PR has the label work-in-progress, please revisit to see if you still need this, please close if not

aojea committed:
fix e2e addnig contexts@

Change-Id: I971b52a6c33de6ff9caf36d4961254fd3854f466

aojea commented:
/close

I'm going to split it in multiple PRs, this is hard to review and/or iterate on
